crates/pallets/starknet/src/lib.rs

```diff
@@ -1092,64 +1090,82 @@ impl<T: Config> Pallet<T> {
     }
 
     /// Estimate the fee associated with transaction
-    pub fn estimate_fee(transaction: UserTransaction, is_query: bool) -> Result<(u64, u64), DispatchError> {
+    pub fn estimate_fee(transactions: Vec<UserTransaction>) -> Result<Vec<(u64, u64)>, DispatchError> {
         let chain_id = Self::chain_id();
 
-        fn execute_tx_and_rollback<S: State + StateChanges + FeeConfig>(
-            tx: impl Execute,
-            state: &mut S,
+        fn execute_txs_and_rollback<T: pallet::Config>(
+            txs: Vec<UserTransaction>,
             block_context: &BlockContext,
             disable_nonce_validation: bool,
-        ) -> TransactionExecutionResult<TransactionExecutionInfo> {
-            // TODO: initialization can probably be skiped by using mem::MaybeUninit
-            let mut execution_result = Ok(Default::default());
+            chain_id: Felt252Wrapper,
+        ) -> Vec<TransactionExecutionResult<TransactionExecutionInfo>> {
+            let mut execution_results = vec![];
             let _: Result<_, DispatchError> = storage::transactional::with_transaction(|| {
-                execution_result = tx.execute(state, block_context, true, disable_nonce_validation);
+                for tx in txs {
```

Comment:
You should use an iterator and map, and stop iterating at the first error encountered.
`collect::<Result<Value, E>>` allows you to do this.

Here you keep executing every transaction, even after a previous one failed. This is a waste of computation

Reply:
And you just do `exectuction_results = txs.into_iter().map(|tx| {...})` to bring the result outside the closure